pkg/secrets/vault.go

```diff
@@ -2,23 +2,242 @@ package secrets
-	return &VaultDecrypter{secretConfig}
+func getVaultDecrypter(encryptedSecret string) Decrypter {
+	decrypter := NewVaultDecrypter(encryptedSecret)
+	decrypter.vaultConfig = Registry.VaultConfig
+	return decrypter
 }
 
 func (v *VaultDecrypter) Decrypt() (string, error) {
-	return "", fmt.Errorf("vault secrets unsupported")
+	if err := v.ValidateVaultConfig(); err != nil {
+		return "", fmt.Errorf("vault configuration error - %s", err)
+	}
+
+	vaultSecret, err := ParseVaultEncryptedSecret(v.encryptedSecret)
+	if err != nil {
+		return "", fmt.Errorf("error parsing vault secret syntax - %s", err)
+	}
+
+	if v.vaultConfig.Token == "" {
+		token, err := v.FetchVaultToken()
+		if err != nil {
+			return "", fmt.Errorf("error fetching vault token - %s", err)
+		}
+		v.vaultConfig.Token = token
+	}
+
+	secret, err := v.FetchSecret(vaultSecret)
+	if err != nil {
+		// get new token and retry in case our saved token is no longer valid
+		return v.RetryFetchSecret(vaultSecret)
+	}
+	return secret, nil
+}
+
+func (v *VaultDecrypter) ValidateVaultConfig() error {
+	if (VaultConfig{}) == v.vaultConfig {
+		return fmt.Errorf("vault secrets not configured in service profile yaml")
+	}
+	if v.vaultConfig.Enabled == false {
+		return fmt.Errorf("vault secrets disabled")
+	}
+	if v.vaultConfig.AuthMethod == "" {
+		return fmt.Errorf("auth method required")
+	}
+	if v.vaultConfig.Url == "" {
+		return fmt.Errorf("vault url required")
+	}
+	return nil
+}
+
+func ParseVaultEncryptedSecret(encryptedSecret string) (VaultSecret, error) {
+	var vaultSecret VaultSecret
+	configs := strings.Split(encryptedSecret, "!")
+	if len(configs) < 2 {
+		return VaultSecret{}, fmt.Errorf("illegal format: %q", encryptedSecret)
+	}
+	for _, element := range configs {
+		kv := strings.Split(element, ":")
+		if len(kv) < 2 {
+			return VaultSecret{}, fmt.Errorf("illegal format for key-value pair in %q: %s",
+				encryptedSecret, element)
+		}
+		switch kv[0] {
+		case "encrypted":
+			// do nothing
+		case "e":
+			vaultSecret.engine = kv[1]
+		case "n":
+			vaultSecret.path = kv[1]
+		case "k":
+			vaultSecret.key = kv[1]
+		case "b":
+			vaultSecret.base64Encoded = kv[1]
+		default:
+			return VaultSecret{}, fmt.Errorf("invalid key in %q: %s", encryptedSecret, kv[0])
+		}
+	}
+	return vaultSecret, nil
+}
+
+func (v *VaultDecrypter) FetchVaultToken() (string, error) {
+	if (VaultConfig{}) == v.vaultConfig {
+		return "", fmt.Errorf("vault secrets not configured in service profile yaml")
+	}
+	if v.vaultConfig.AuthMethod == "TOKEN" {
+		token := os.Getenv("VAULT_TOKEN")
+		if token != "" {
+			return token, nil
+		} else {
+			return "", fmt.Errorf("VAULT_TOKEN environment variable not set")
+		}
+	} else if v.vaultConfig.AuthMethod == "KUBERNETES" {
+		if v.vaultConfig.Path == "" || v.vaultConfig.Role == "" {
+			return "", fmt.Errorf("path and role both required for Kubernetes auth method")
+		}
+		return v.FetchServiceAccountToken()
```

Comment:
Is there no Vault SDK method to handle kubernetes service account login? That would be surprising.

Reply:
I couldn't find one?